File: hw01/problem02.py
import heapq
import math

class Node:
    def __init__(self, id, a, b, c, d, price):
        self.id = id
        self.a = a
        self.b = b
        self.c = c
        self.d = d
        self.price = price
    
    def __repr__(self):
        return f"id: {self.id} : {self.a} {self.b} {self.c} {self.d} {self.price}"
    
    def l2_distance(self):
        return math.sqrt((self.a - BASE.a) ** 2 +
                         (self.b - BASE.b) ** 2 +
                         (self.c - BASE.c) ** 2 +
                         (self.d - BASE.d) ** 2)

    def __lt__(self, other):
        return self.l2_distance() < other.l2_distance()

BASE = Node(-1, 6, 200, 5, 30, -1)

nodes = [
    Node(1, 2, 200, 4, 27, 30000),
    Node(2, 5, 150, 3, 35, 20000),
    Node(3, 3, 180, 4, 25, 25000),
    Node(4, 1, 230, 2, 10, 21000),
    Node(5, 5, 180, 5, 40, 38000),
    Node(6, 4, 210, 3, 30, 31000)
]

pq = []
for node in nodes:
    heapq.heappush(pq, node)

price = 0
weight = 0

for _ in range(5):
    now = heapq.heappop(pq)
    print(now)
    temp_weight = math.exp(-now.l2_distance())
    print(f"tempWeight : {temp_weight}")
    price += temp_weight * now.price
    weight += temp_weight

print(f"price : {price / weight}")

File: hw01/problem02.cpp
#include <iostream>
#include <queue>
#include <vector>
#include <unordered_map>
#include <cmath>

const int N=3;

struct Node{
    int id;
    int a;
    int b;
    int c;
    int d;
    int price;
};

std::ostream& operator<<(std::ostream& out, const Node& node) {
    out <<"id: "<<node.id<<" : "<< node.a << " " << node.b << " " << node.c << " " << node.d << " " << node.price;
    return out;
}

const Node BASE={-1,6,200,5,30,-1};

const double calcL2Distance(const Node& n) {
    return std::sqrt((n.a - BASE.a) * (n.a - BASE.a) +
                     (n.b - BASE.b) * (n.b - BASE.b) +
                     (n.c - BASE.c) * (n.c - BASE.c) +
                     (n.d - BASE.d) * (n.d - BASE.d));
}


struct Compare {
    bool operator()(const Node& n1, const Node& n2) {
        return calcL2Distance(n1) > calcL2Distance(n2);
    }
};



int main(){
    std::vector<Node> v;
    v.push_back({1,2,200,4,27,30000});
    v.push_back({2,5,150,3,35,20000});
    v.push_back({3,3,180,4,25,25000});
    v.push_back({4,1,230,2,10,21000});
    v.push_back({5,5,180,5,40,38000});
    v.push_back({6,4,210,3,30,31000});


    std::priority_queue<Node,std::vector<Node>,Compare> pq(v.begin(),v.end());

    double price=0;
    double weight=0;

    for(int i=0;i<5;++i){
        const Node now(pq.top());
        pq.pop();

        std::cout<<now<<'\n';
        const double tempWeight=exp(-calcL2Distance(now));
        std::cout<<"tempWeight : "<<tempWeight<<'\n';
        price+=tempWeight*now.price;
        weight+=tempWeight;
    }

    std::cout<<"price : "<<price/weight<<'\n';
}